### DISCUSSION TITLE
How does casadi deal with abs(x) in the background?

### DISCUSSION THREAD

**Discussion by Jimmy0212Fan** [casadi/casadi#3228] [Q&A]
Hi,
I want to ask how casadi deal with the abs() of the constraints when solving a NLP. Does casadi use a kind of approximation method? if yes, what kinds of approximation mehod? Please feel free to answer! 
Regards,
Jiayu

**Comment by jaeandersson** [comment]
CasADi will calculate the derivative which is well-defined almost everywhere. Since the constraints won't be continuously differentiable, the solver may fail. It's usually a good idea to try to reformulate the abs() unless you know that the sign won't change during optimization.

**Reply by Jimmy0212Fan** [reply]
Thanks very much for the quick replying.
But I still have puzzles, please see the following example. 

This NLP problem EXIT with "Optimal Solution Found".  The solving begins with x_0 =-5000, thus abs(x+3000)<0;  when x=-2 (optimal), thus abs(x+3000)>0.  During the solving, the sign of abs() changes but CasAdi can still deal with abs() successfully.  I don't know why. 
Additionally, after running, I checked the g in workspace, abs() is denoted by fabs(). Does CasAdi deal with fabs() in the background?

%% Matlab
clc;clear;
% Symbols/expressions
x = casadi.MX.sym('x');
y = casadi.MX.sym('y');
f = (x+2)^2;
g = (1-abs(x+3000))^3-y; % g

nlp = struct;            % NLP declaration
nlp.x = [x;y];         % decision vars
nlp.f = f;               % objective
nlp.g = g;               % constraints

% Create solver instance
F = casadi.nlpsol('F','ipopt',nlp);

% Solve the problem using a guess
sol =F('x0',[-5000 -3.0 ],'ubg',0,'lbg',0,'lbx',[-inf -inf ],'ubx',[inf inf]);
wopt = full(sol.x);
%%%

**Reply by jaeandersson** [reply]
CasADi can handle the sign change. The derivative of abs(x) is essentially sign(x). The problem is that the NLP solver which is ipopt here, may not converge or may converge much slower. Often ipopt will work even if the constraints are nonsmooth, but bets are off

**Reply by Jimmy0212Fan** [reply]
OK, I get it. Thanks a lot!

**Reply by Jimmy0212Fan** [reply]
hi, apart from the abs(x), I have one more question about the max and min function. Could you please explain how casadi deal with max(0,x) and min(0,x) in the background, how to calculate the derivative of  max(0,x) and min(0,x)? Thanks in advance!

**Reply by jaeandersson** [reply]
It's the same thing. There is an analytic differentiation rule and the derivatives are well-defined almost everywhere. For example the partial derivative of min(x,y) with respect to x is (x<y) almost everywhere

**Reply by Jimmy0212Fan** [reply]
Thanks for the quick reply! 
So the derivative of min(x,0) is 1 (x<0) and 0 (x>0). But what about the derivative of min(x,0) when x=0?  Is it equal to 0 or 1?

**Comment by jgillis** [comment]
> Thanks for the quick reply! So the derivative of min(x,0) is 1 (x<0) and 0 (x>0). But what about the derivative of min(x,0) when x=0? Is it equal to 0 or 1?

In 3.6, it changed to 0.5 iirc think: https://github.com/casadi/casadi/blob/b3ccf30fc2072933a5f3e797989554525a5c2dc9/casadi/core/calculus.hpp#L883
